Mathlib/ModelTheory/Algebra/Field/IsAlgClosed.lean

```diff
@@ -0,0 +1,285 @@
+algebraically closed field `K` which does not have a `Language.ring.Structure` instance,
+you must introduce the local instance `compatibleRingOfRing K`. Theorems whose statement requires
+both a `Language.ring.Structure` instance and a `Field` instance will all be stated with the
+assumption `Field K`, `CharP K p`, `IsAlgClosed K` and `CompatibleRing K` and there are instances
+defined saying that these assumption imply `Theory.field.Model K` and `(Theory.ACF p).Model K`
+
+## References
+The first order theory of algebraically closed fields, along with the Lefschetz Principle and
+the Ax-Grothendieck Theorem were first formalized in Lean 3 by Joseph Hua
+[here](https://github.com/Jlh18/ModelTheoryInLean8) with the master's thesis
+[here](https://github.com/Jlh18/ModelTheory8Report)
+
+-/
+
+variable {K : Type*}
+
+namespace FirstOrder
+
+namespace Field
+
+open Ring FreeCommRing BigOperators Polynomial Language
+
+/-- A generic monic polynomial of degree `n` as an element of the
+free commutative ring in `n+1` variables, with a variable for each
+of the `n` non-leading coefficients of the polynomial and one variable (`Fin.last n`)
+for `X`.  -/
+def genericMonicPoly (n : ℕ) : FreeCommRing (Fin (n + 1)) :=
+  of (Fin.last _) ^ n + ∑ i : Fin n, of i.castSucc * of (Fin.last _) ^ (i : ℕ)
+
+/-- Monic polynomials of degree `n` over `K` are equivalent to functions `Fin n → K`  -/
+noncomputable def monicPolyEquivFin [CommRing K] [Nontrivial K] (n : ℕ) :
+    { p : K[X] // p.Monic ∧ p.natDegree = n } ≃ (Fin n → K) :=
+  { toFun := fun p i => p.1.coeff i
+    invFun := fun v =>
+      let p := Polynomial.ofFinsupp <|
+        Finsupp.ofSupportFinite
+          (fun i =>
+            if h : i < n then v ⟨i, h⟩
+            else if i = n then 1 else 0)
+          (Set.Finite.subset (Set.finite_le_nat n) <| by
+              intro i
+              simp
+              split_ifs <;> simp_all [le_iff_lt_or_eq])
+      have hpn : p.natDegree = n := by
+        refine le_antisymm ?_ ?_
+        · refine natDegree_le_iff_coeff_eq_zero.2 ?_
+          intro i hi
+          simp [not_lt_of_gt hi, ne_of_gt hi, Finsupp.ofSupportFinite]
+        · refine le_natDegree_of_ne_zero ?_
+          simp [Finsupp.ofSupportFinite]
+      have hpm : p.Monic := by
+        rw [Monic.def, leadingCoeff, hpn]
+        simp [Finsupp.ofSupportFinite]
+      ⟨p, hpm, hpn⟩
+    left_inv := by
+      intro p
+      ext i
+      simp only [ne_eq, Finset.mem_range, dite_eq_ite, coeff_ofFinsupp,
+        Finsupp.coe_mk, ite_eq_left_iff, not_lt, Finsupp.ofSupportFinite]
+      intro hni
+      cases lt_or_eq_of_le hni with
+      | inr =>
+        subst n
+        have := p.2.1
+        simp [Monic.def, leadingCoeff, p.2.2] at this
+        simp [this]
+      | inl h =>
+        rw [eq_comm, if_neg (ne_of_gt h)]
+        exact coeff_eq_zero_of_natDegree_lt (p.2.2.symm ▸ h)
+    right_inv := fun _ => by simp [Finsupp.ofSupportFinite] }
```

Comment:
This looks quite complicated for something so simple. @eric-wieser, do we not have ways to restrict a `Finsupp` to a superset of its support, so that this equivalence can be obtained as the composition of this and the equivalence `(ι → α) ≃ (ι →₀ α)` for finite `ι`?